### PR TITLE
Fix watchDmChannel may be out-of-date after compaction issue

### PR DESCRIPTION
**Pull request by wayblink**
related: #18619

Signed-off-by: wayblink <anyang.wang@zilliz.com>

### CODE REVIEW

**Comment by wayblink** [comment]
rerun ut

**Comment by wayblink** [comment]
rerun ut

File: internal/querycoord/task_util.go [hunk 1]
```diff
@@ -34,25 +37,74 @@ func generateFullWatchDmChannelsRequest(broker *globalMetaBroker, request *query
...
+	// refresh the vchannels and segmentInfos needed.
+	retryFunc := func() error {
+		newVChannels := make([]*datapb.VchannelInfo, 0)
+		newSegmentIds := make([]int64, 0)
+
+		newVChannelDict := make(map[string]bool, 0)
+		for _, partitionID := range request.GetLoadMeta().GetPartitionIDs() {
+			partitionVChannels, _, err := broker.getRecoveryInfo(ctx, request.GetCollectionID(), partitionID)
+			if err != nil {
+				log.Error("GetRecoveryInfo failed", zap.Error(err))
```
Comment (soothing-rain) [inline internal/querycoord/task_util.go:56]: NIT:

"GetRecoveryInfo failed" -> "GetRecoveryInfo failed, retrying..."

File: internal/querycoord/task_util.go [hunk 2]
```diff
@@ -34,25 +37,74 @@ func generateFullWatchDmChannelsRequest(broker *globalMetaBroker, request *query
...
+		}
+
+		for _, vChannel := range newVChannels {
+			newSegmentIds = append(newSegmentIds, vChannel.FlushedSegmentIds...)
+			newSegmentIds = append(newSegmentIds, vChannel.UnflushedSegmentIds...)
+			newSegmentIds = append(newSegmentIds, vChannel.DroppedSegmentIds...)
+		}
+		newSegmentInfos, err := broker.getDataSegmentInfosByIDs(ctx, newSegmentIds)
+		if err != nil {
+			log.Error("Get Vchannel SegmentInfos failed", zap.Error(err))
```
Comment (soothing-rain) [inline internal/querycoord/task_util.go:74]: ditto

**Comment by wayblink** [comment]
rerun ut

File: internal/querycoord/task_scheduler.go [hunk 3]
```diff
@@ -393,7 +393,7 @@ func (scheduler *TaskScheduler) unmarshalTask(taskID UniqueID, t string) (task,
 		if err != nil {
 			return nil, err
 		}
-		fullReq, err := generateFullWatchDmChannelsRequest(scheduler.broker, &req)
+		fullReq, err := generateFullWatchDmChannelsRequest(scheduler.ctx, scheduler.broker, &req)
 		if err != nil {
```
Comment (congqixia) [inline internal/querycoord/task_scheduler.go:397]: `scheduler.ctx` or `task.traceCtx()`?

**Review by congqixia — COMMENTED** [review]
overall lgtm
Please check the ctx instance we should use when calling `generateFullWatchDmChannelsRequest`

File: internal/querycoord/task_scheduler.go (hunk 3 above)
Comment (wayblink) [reply internal/querycoord/task_scheduler.go:397]: nice catch, I think task.traceCtx() is better

**Pull request merged**